src/node_file.cc

```diff
@@ -3391,6 +3416,7 @@ static void CreatePerIsolateProperties(IsolateData* isolate_data,
   SetMethod(isolate, target, "fchmod", FChmod);
 
   SetMethod(isolate, target, "chown", Chown);
+  SetMethodNoSideEffect(isolate, target, "chownSync", ChownSync);
```

Comment:
I'm a bit surprised. Isn't "chown" a side-effect by itself?

Reply:
It doesn't mutate the environment at all when running on `sync` path.

Reply:
@anonrig But it mutates the user's file system, doesn't it? That's a pretty big side effect.

Reply:
Yes, but it doesn't trigger JS-observable side effects.

According to the documentation, ref: https://github.com/v8/v8/blob/219b0dc7379363b1a8df7e144c78ca5dc09a8d69/include/v8-object.h#L204

```
/**
 * Options for marking whether callbacks may trigger JS-observable side effects.
 * Side-effect-free callbacks are allowlisted during debug evaluation with
 * throwOnSideEffect. It applies when calling a Function, FunctionTemplate,
 * or an Accessor callback. For Interceptors, please see
 * PropertyHandlerFlags's kHasNoSideEffect.
 * Callbacks that only cause side effects to the receiver are allowlisted if
 * invoked on receiver objects that are created within the same debug-evaluate
 * call, as these objects are temporary and the side effect does not escape.
 */
```

Any file system changes does not create JS-observable side effects.

Reply:
Huh, I must have misunderstood that flag. I assumed that any side effect would count.

Reply:
>  I assumed that any side effect would count.

I think that's correct. The purpose of this flag basically comes down to "when you type the prefix of this call in the REPL, can the REPL eagerly evaluate it and tell you the result before you type enter" and I think for chown, the answer is no. You don't want the chown to happen before you type enter.

Reply:
@joyeecheung how does kHasSideEffectToReceiver fit into this equation?

Reply:
Applied the changes. Resolving the conversation.